network/p2p/inspector/validation/control_message_validation_inspector_test.go

```diff
@@ -683,6 +683,82 @@ func TestControlMessageValidationInspector_processInspectRPCReq(t *testing.T) {
+				unittest.P2PRPCIHaveFixture(&malformedTopic, unittest.GenerateRandomStringsWithLen(10, 10)...),
+				unittest.P2PRPCIHaveFixture(&invalidSporkIDTopic, unittest.GenerateRandomStringsWithLen(10, 10)...),
+				unittest.P2PRPCIHaveFixture(&validTopic, duplicateMessageID, duplicateMessageID)}
+			invalidIWants := []*pubsub_pb.ControlIWant{unittest.P2PRPCIWantFixture(duplicateMessageID, duplicateMessageID)}
+
+			graftsReq := unittest.P2PRPCFixture(unittest.WithGrafts(invalidGrafts...))
+			prunesReq := unittest.P2PRPCFixture(unittest.WithPrunes(invalidPrunes...))
+			ihavesReq := unittest.P2PRPCFixture(unittest.WithIHaves(invalidIHaves...))
+			iwantsReq := unittest.P2PRPCFixture(unittest.WithIWants(invalidIWants...))
+
+			from := unittest.PeerIdFixture(t)
+			ensureIsInvalidTopicErr := func(errors p2p.InvCtrlMsgErrs) {
+				for _, err := range errors {
+					require.True(t, channels.IsInvalidTopicErr(err.Err))
+				}
+			}
+			distributor.On("Distribute", mock.AnythingOfType("*p2p.InvCtrlMsgNotif")).Return(nil).Times(4).Run(func(args mock.Arguments) {
+				notification, ok := args[0].(*p2p.InvCtrlMsgNotif)
+				require.True(t, ok)
+				require.Equal(t, from, notification.PeerID)
+				switch notification.MsgType {
+				case p2pmsg.CtrlMsgGraft:
+					require.Equal(t, len(invalidGrafts), notification.Errors.Len())
+					ensureIsInvalidTopicErr(notification.Errors)
+				case p2pmsg.CtrlMsgPrune:
+					require.Equal(t, len(invalidPrunes), notification.Errors.Len())
+					ensureIsInvalidTopicErr(notification.Errors)
+				case p2pmsg.CtrlMsgIHave:
+					require.Equal(t, len(invalidIHaves), notification.Errors.Len())
+					for _, err := range notification.Errors {
+						switch {
+						case channels.IsInvalidTopicErr(err.Err):
+							continue
+						case validation.IsDuplicateMessageIDErr(err.Err):
+							continue
+						default:
+							require.Fail(t, fmt.Sprintf("unexpected error encountered %s for iHave control message", err.Err))
+						}
+					}
+				case p2pmsg.CtrlMsgIWant:
+					require.Equal(t, len(invalidIWants), notification.Errors.Len())
+					require.True(t, validation.IsIWantDuplicateMsgIDThresholdErr(notification.Errors[0].Err))
+				default:
+					require.Fail(t, fmt.Sprintf("unexpected control message type: %s", notification.MsgType))
+				}
+			})
+			rpcTracker.On("LastHighestIHaveRPCSize").Return(int64(100)).Maybe()
+			rpcTracker.On("WasIHaveRPCSent", mock.AnythingOfType("string")).Return(true).Run(func(args mock.Arguments) {
+				id, ok := args[0].(string)
+				require.True(t, ok)
+				require.Contains(t, duplicateMessageID, id)
+			})
+			inspector.Start(signalerCtx)
+
+			require.NoError(t, inspector.Inspect(from, graftsReq))
+			require.NoError(t, inspector.Inspect(from, prunesReq))
+			require.NoError(t, inspector.Inspect(from, ihavesReq))
+			require.NoError(t, inspector.Inspect(from, iwantsReq))
+			// sleep for 1 second to ensure rpc's is processed
+			time.Sleep(time.Second)
```

Comment:
⚠️ this can cause flakey behavior for the test specially over CI. Please use a channel or wait group when mocking the "Distribute" method of the `distributor` and use a timeout-based wait on that channel or wait group.

Reply:
https://github.com/onflow/flow-go/pull/4978/commits/850fb571e8c39f6554b5c6b51bf4a582dfbf403a